File: Problems/Rank Sales Leads/rank_leads_xgb.R
# gradient boosting model using xgboost

# Notes about this model:
# ...

options(scipen=20, digits=4)

# Load packages
library(data.table)
library(stringr)
library(Matrix)
library(xgboost)
library(pROC)

#======================================================================================================
# Load Data (Assumes your current working directory is the Rank Sales Leads problem directory)

train <- fread("_Data/train.csv")
test <- fread("_Data/test.csv")

#======================================================================================================
# Really quick and dirty analysis

# View the raw data
train

# Some thoughts & ideas come to mind
# - CompanyName could be useful in theory, but we're not going to deal with the complexities of NLP feature engineering here...
# - Phone number probably isn't useful... but maybe AreaCode is since it should separate the samples by geography => demographics
# - It probably makes sense to order the Contact values: general line < other < manager < owner (this affects the model's construction)
# - Website (like PhoneNumber) is too unique to be useful, but maybe businesses that use .net and .org are less likely to buy our software

# Now let's see what our overall hit ratio is
train[, list(Samples=.N, Sales=sum(Sale), HitRatio = sum(Sale)/.N)]  # .35

#======================================================================================================
# Feature engineering and transforming the training dataset

# Some things to keep in mind
# - We have to deal with missing values (NAs). We could impute something into their place, but this is likely to degrade
#   the model's performace. NAs here have special meaning. E.g. FacebookLikes = NA means the company does not have a facebook.
#   This has very different implications than, "The company has a facebook, but we didn't bother to track how many Likes it has"
# - randomForest only accepts numeric and factor values and it can't handle NAs. So we need to convert categorical features to factors
#   and change NAs to some specific value, e.g. "NA_Val", or -1 for numeric features. Also need to convert Sale from logical to factor
# - randomForest can only handle factors with up to 53 unique levels. We won't run into this issue since out dataset is small, but for
#   illustration we'll create a "catch all" group for TypeOfBusiness to use for uncommon business types
# - The test set could have a TypeOfBusiness not seen in the training set. Need to prepare for that!

#--------------------------------------------------
# Create some row indexes to help us populate sparse matricies

train[, RowIdx := .I]
test[, RowIdx := .I]

#--------------------------------------------------
# TypeOfBusiness

# Generate a map to map the values in TypeOfBusiness to a sparse matrix
tobMap <- train[, list(Samples=.N), by=TypeOfBusiness]
tobMap[, TobIdx := .I]

# train
train[tobMap, TobIdx := TobIdx, on="TypeOfBusiness"]
trainTOBSparseM <- sparseMatrix(i=train$RowIdx, j=train$TobIdx, x=1, dimnames=list(NULL, tobMap$TypeOfBusiness))  # build a sparse matrix

# test
test[tobMap, TobIdx := TobIdx, on="TypeOfBusiness"]
testTOBSparseM <- sparseMatrix(i=test[!is.na(TobIdx)]$RowIdx, j=test[!is.na(TobIdx)]$TobIdx, x=1, dims=c(nrow(test), nrow(tobMap)), dimnames=list(NULL, tobMap$TypeOfBusiness))

#--------------------------------------------------
# Contact (convert to numeric, 1-4)

# In this case, we know all the possible contact types
contacts <- c("general line", "other", "manager", "owner")  # Note the order of the elements
train[, Contact := match(Contact, contacts)]
test[, Contact := match(Contact, contacts)]

#--------------------------------------------------
# AreaCode

train[, AreaCode := substr(PhoneNumber, 1, 3)]  # train
test[, AreaCode := substr(PhoneNumber, 1, 3)]  # test

# Generate a map to map the values in AreaCode to a sparse matrix
areacodeMap <- train[, list(Samples=.N), by=AreaCode]
areacodeMap[, AreaCodeIdx := .I]

# train
train[areacodeMap, AreaCodeIdx := AreaCodeIdx, on="AreaCode"]
trainAreaCodeSparseM <- sparseMatrix(i=train$RowIdx, j=train$AreaCodeIdx, x=1, dimnames=list(NULL, areacodeMap$AreaCode))  # build a sparse matrix

# test
test[areacodeMap, AreaCodeIdx := AreaCodeIdx, on="AreaCode"]
testAreaCodeSparseM <- sparseMatrix(i=test[!is.na(AreaCodeIdx)]$RowIdx, j=test[!is.na(AreaCodeIdx)]$AreaCodeIdx, x=1, dims=c(nrow(test), nrow(areacodeMap)), dimnames=list(NULL, areacodeMap$AreaCode))

#--------------------------------------------------
# Website Extension  (note: str_detect comes from the stringr package. Thanks Hadley)

extensions <- c("com", "net", "org", "other", "none")

# train
train[is.na(Website), WebsiteExtension := "none"]
train[str_detect(Website, ".com"), WebsiteExtension := "com"]
train[str_detect(Website, ".net"), WebsiteExtension := "net"]
train[str_detect(Website, ".org"), WebsiteExtension := "org"]
train[is.na(WebsiteExtension), WebsiteExtension := "other"]

# test
test[is.na(Website), WebsiteExtension := "none"]
test[str_detect(Website, ".com"), WebsiteExtension := "com"]
test[str_detect(Website, ".net"), WebsiteExtension := "net"]
test[str_detect(Website, ".org"), WebsiteExtension := "org"]
test[is.na(WebsiteExtension), WebsiteExtension := "other"]

# Generate a map to map the values in WebsiteExtension to a sparse matrix
extensionMap <- train[, list(Samples=.N), by=WebsiteExtension]
extensionMap[, ExtensionIdx := .I]

# train
train[extensionMap, ExtensionIdx := ExtensionIdx, on="WebsiteExtension"]
trainExtensionSparseM <- sparseMatrix(i=train$RowIdx, j=train$ExtensionIdx, x=1, dimnames=list(NULL, extensionMap$WebsiteExtension))  # build a sparse matrix

# test
test[extensionMap, ExtensionIdx := ExtensionIdx, on="WebsiteExtension"]
testExtensionSparseM <- sparseMatrix(i=test[!is.na(ExtensionIdx)]$RowIdx, j=test[!is.na(ExtensionIdx)]$ExtensionIdx, x=1, dims=c(nrow(test), nrow(extensionMap)), dimnames=list(NULL, extensionMap$WebsiteExtension))

#--------------------------------------------------
# FacebookLikes

# Note -1L is type integer vs -1 which is type double
train[is.na(FacebookLikes), FacebookLikes := -1L]  # train
test[is.na(FacebookLikes), FacebookLikes := -1L]  # test

#--------------------------------------------------
# TwitterFollowers

train[is.na(TwitterFollowers), TwitterFollowers := -1L]  # train
test[is.na(TwitterFollowers), TwitterFollowers := -1L]  # test

#--------------------------------------------------
# Sale (the target variable)

# Convert to numeric for xgboost
train[, Sale := Sale*1]

#--------------------------------------------------
# Combine training features into a single sparse matrix

# Insert the non sparse features into a dgCMatrix
trainNonSparseFeats <- Matrix(as.matrix(train[, list(Contact, FacebookLikes, TwitterFollowers)]), sparse=TRUE)
testNonSparseFeats <- Matrix(as.matrix(test[, list(Contact, FacebookLikes, TwitterFollowers)]), sparse=TRUE)

# Combine all the sparse matrices into one big sparse matrix
trainM <- do.call(cBind, list(trainNonSparse, trainTOBSparseM, trainAreaCodeSparseM, trainExtensionSparseM))
testM <- do.call(cBind, list(testNonSparse, testTOBSparseM, testAreaCodeSparseM, testExtensionSparseM))

# Extract the column names of the matrix into a vector called 'features'
features <- c(dimnames(trainM)[[2]])

#======================================================================================================
# XGBoost Model

set.seed(2016)
boostingParams = list(objective="binary:logistic", eta=0.01, max.depth=3, eval_metric="auc", colsample_bytree=.75)
bst <- xgboost(params=boostingParams, data=trainM, label=as.matrix(train$Sale)*1, nrounds=10)

#--------------------------------------------------
# Check the feature importances

xgb.importance(bst, feature_names=features)

#======================================================================================================
# Make some predictions on the test set & evaluate the results

# Make Predictions
test[, ProbSale := predict(bst, newdata=testM)]

#--------------------------------------------------
# Rank the predictions from most likely to least likely

setorder(test, -ProbSale)
test[, ProbSaleRk := .I]

#--------------------------------------------------
# Take a look

test[, list(ProbSaleRk, CompanyName, ProbSale, Sale)]  # Looks perty good!

#--------------------------------------------------
# Let's evaluate the results using area under the ROC curve using the pROC package

rocCurve <- roc(response=test$Sale, predictor=test$ProbSale, direction="<")
rocCurve$auc  # 0.844
plot(rocCurve)

File: Problems/Rank Sales Leads/rank_leads_xgb.py
# Gradient Boosting model using XGBoost

# Notes about this model:
# XGBoost requires a single matrix of numeric values as input. This means non ordered categorical features need to be
# one-hot-encoded. This can result in a really large sparse matrix, but XGBoost allows for the input matrix to be represented
# in compressed format via dgCMatrix, so we'll utilize that option here

# Imports
import os
import numpy as np
import pandas as pd
from scipy import sparse
import xgboost as xgb
from sklearn.metrics import roc_auc_score

# Display Settings
pd.set_option('display.max_rows', 10)
pd.set_option('display.max_columns', 20)
pd.set_option('display.width', 190)

# # Set working directory
# os.chdir("/Path/To/Rank Sales Leads")

#======================================================================================================
# Load Data (Assumes your current working directory is the Rank Sales Leads problem directory)

train = pd.read_csv("_Data/train.csv", dtype={'PhoneNumber':str})
test = pd.read_csv("_Data/test.csv", dtype={'PhoneNumber':str})

#======================================================================================================
# Really quick and dirty analysis

# View the raw data
train

# Some thoughts & ideas come to mind
# - CompanyName could be useful in theory, but we're not going to deal with the complexities of NLP feature engineering here...
# - Phone number probably isn't useful... but maybe AreaCode is since it should separate the samples by geography => demographics
# - It probably makes sense to order the Contact values: general line < other < manager < owner (this affects the model's construction)
# - Website (like PhoneNumber) is too unique to be useful, but maybe businesses that use .net and .org are less likely to buy our software

# Now let's see what our overall hit ratio is
train.Sale.sum()/train.shape[0]  # .35

#======================================================================================================
# Feature engineering and transforming the training dataset

# Some things to keep in mind
# - Need to build a sparse matrix for each non-ordered categorical feature: TypeOfBusiness, AreaCode, and WebsiteExtension
# - Need to concatenate all features into one big sparse matrix
# - XGBoost has smart handling for NA values, so we don't need to impute values for NA in FacebookLikes and TwitterFollowers

#--------------------------------------------------
# TypeOfBusiness

# Convert NAs to "NA_Val"
train.TypeOfBusiness.fillna("NA_Val", inplace=True)
test.TypeOfBusiness.fillna("NA_Val", inplace=True)

# Generate a map to map the values in TypeOfBusiness to a sparse matrix
tobMap = pd.DataFrame({'TypeOfBusiness': train.TypeOfBusiness.unique()})
tobMap['TOBColIdx'] = np.arange(tobMap.shape[0])

# Build sparse matrix of train values
train = train.merge(tobMap, on='TypeOfBusiness', how='left')
trainTOBSparseM = sparse.csr_matrix((np.repeat(1, train.shape[0]), (train.index.values, train.TOBColIdx.values)))

# Build sparse matrix of test values
test = test.merge(tobMap, on='TypeOfBusiness', how='left')
testTOBSparseM = sparse.csr_matrix((np.repeat(1, sum(test.TOBColIdx.notnull())), (test[test.TOBColIdx.notnull()].index.values, test[test.TOBColIdx.notnull()].TOBColIdx.values)), [test.shape[0], tobMap.shape[0]])

#--------------------------------------------------
# AreaCode

train['AreaCode'] = train.PhoneNumber.str[:3]
test['AreaCode'] = test.PhoneNumber.str[:3]

# Generate a map to map the values in AreaCode to a sparse matrix
areacodeMap = pd.DataFrame({'AreaCode': train.AreaCode.unique()})
areacodeMap['AreaCodeColIdx'] = np.arange(areacodeMap.shape[0])

# Build sparse matrix of train values
train = train.merge(areacodeMap, on='AreaCode', how='left')
trainAreaCodeSparseM = sparse.csr_matrix((np.repeat(1, train.shape[0]), (train.index.values, train.AreaCodeColIdx.values)))

# Build sparse matrix of test values
test = test.merge(areacodeMap, on='AreaCode', how='left')
testAreaCodeSparseM = sparse.csr_matrix((np.repeat(1, sum(test.AreaCodeColIdx.notnull())), (test[test.AreaCodeColIdx.notnull()].index.values, test[test.AreaCodeColIdx.notnull()].AreaCodeColIdx.values)), [test.shape[0], areacodeMap.shape[0]])

#--------------------------------------------------
# Website Extension

# train
train.loc[train.Website.isnull(), 'WebsiteExtension'] = 'none'
train.loc[train.Website.str.contains('com').replace(np.nan, False), 'WebsiteExtension'] = 'com'
train.loc[train.Website.str.contains('org').replace(np.nan, False), 'WebsiteExtension'] = 'org'
train.loc[train.Website.str.contains('net').replace(np.nan, False), 'WebsiteExtension'] = 'net'
train.loc[train.WebsiteExtension.isnull(), 'WebsiteExtension'] = 'other'

# test
test.loc[test.Website.isnull(), 'WebsiteExtension'] = 'none'
test.loc[test.Website.str.contains('com').replace(np.nan, False), 'WebsiteExtension'] = 'com'
test.loc[test.Website.str.contains('org').replace(np.nan, False), 'WebsiteExtension'] = 'org'
test.loc[test.Website.str.contains('net').replace(np.nan, False), 'WebsiteExtension'] = 'net'
test.loc[test.WebsiteExtension.isnull(), 'WebsiteExtension'] = 'other'

# Generate a map to map the values in WebsiteExtension to a sparse matrix
extensionMap = pd.DataFrame({'WebsiteExtension': train.WebsiteExtension.unique()})
extensionMap['ExtensionColIdx'] = np.arange(extensionMap.shape[0])

# Build sparse matrix of train values
train = train.merge(extensionMap, on='WebsiteExtension', how='left')
trainExtensionSparseM = sparse.csr_matrix((np.repeat(1, train.shape[0]), (train.index.values, train.ExtensionColIdx.values)))

# Build sparse matrix of test values
test = test.merge(extensionMap, on='WebsiteExtension', how='left')
testExtensionSparseM = sparse.csr_matrix((np.repeat(1, sum(test.ExtensionColIdx.notnull())), (test[test.ExtensionColIdx.notnull()].index.values, test[test.ExtensionColIdx.notnull()].ExtensionColIdx.values)), [test.shape[0], extensionMap.shape[0]])

#--------------------------------------------------
# Contact (convert to numeric, 1-4)

# In this case, we know all the possible contact types
contacts = ["general line", "other", "manager", "owner"]  # Note the order of the elements
train['Contact'] = pd.Categorical(train.Contact, categories=contacts, ordered=True).codes
test['Contact'] = pd.Categorical(test.Contact, categories=contacts, ordered=True).codes

#--------------------------------------------------
# Combine training features into a single sparse matrix and then convert to xgb.DMatrix type

# Insert the non sparse features into a sparse matrix
trainNonSparseFeats = sparse.csr_matrix(train[['Contact', 'FacebookLikes', 'TwitterFollowers']])
testNonSparseFeats = sparse.csr_matrix(test[['Contact', 'FacebookLikes', 'TwitterFollowers']])

# Concatenate all sparse matrices
trainM = sparse.hstack([trainNonSparseFeats, trainTOBSparseM, trainAreaCodeSparseM, trainExtensionSparseM])
testM = sparse.hstack([testNonSparseFeats, testTOBSparseM, testAreaCodeSparseM, testExtensionSparseM])

# Convert to type DMatrix for xgboost
trainM = xgb.DMatrix(data=trainM, label=train.Sale)
testM = xgb.DMatrix(data=testM, label=test.Sale)

#======================================================================================================
# XGBoost Model

np.random.seed(2016)  # eta=.3, max.depth=10, subsample=.75, colsample_bytree=.75, min_child_weight=1, gamma=0, lambda=0, alpha=0
boosting_params = {'bst:eta':0.3, 'bst:max_depth':10, 'bst:subsample':.75, 'bst:colsample_bytree':.75, 'min_child_weight':1, 'gamma':0, 'lambda':0, 'alpha':0, 'objective':'binary:logistic', 'eval_metric':'auc'}
bst = xgb.train(params=boosting_params, dtrain=trainM, num_boost_round=10)

#======================================================================================================
# Make some predictions on the test set & evaluate the results

test['ProbSale'] = bst.predict(testM)

#--------------------------------------------------
# Rank the predictions from most likely to least likely

test.sort_values('ProbSale', inplace=True, ascending=False)
test['ProbSaleRk'] = np.arange(test.shape[0])

#--------------------------------------------------
# Take a look

test[['ProbSaleRk', 'CompanyName', 'ProbSale', 'Sale']]  # Looks perty good!

#--------------------------------------------------
# Evaluate the results using area under the ROC curve

roc_auc_score(y_true=test.Sale, y_score=test.ProbSale)  # 0.875
